app/upgrades/v16/utils.go

```diff
@@ -0,0 +1,140 @@
+)
+
+const (
+	WEVMOSContractMainnet = "0xD4949664cD82660AaE99bEdc034a0deA8A0bd517"
+	WEVMOSContractTestnet = "0xcc491f589b45d4a3c679016195b3fb87d7848210"
+)
+
+// ConvertERC20Coins converts Native IBC coins from their ERC20 representation
+// to the native representation. This also includes the withdrawal of WEVMOS tokens
+// to EVMOS native tokens.
+func ConvertERC20Coins(
+	ctx sdk.Context,
+	logger log.Logger,
+	accountKeeper authkeeper.AccountKeeper,
+	bankKeeper bankkeeper.Keeper,
+	erc20Keeper erc20keeper.Keeper,
+) error {
+	var wrappedContractAddr common.Address
+
+	if utils.IsMainnet(ctx.ChainID()) {
+		wrappedContractAddr = common.HexToAddress(WEVMOSContractMainnet)
+	} else if utils.IsTestnet(ctx.ChainID()) {
+		wrappedContractAddr = common.HexToAddress(WEVMOSContractTestnet)
+	}
+
+	// iterate over all the accounts and convert the tokens to native coins
+	accountKeeper.IterateAccounts(ctx, func(account authtypes.AccountI) (stop bool) {
+		// FIXME: the accounts created with the keyring are not EthAccounts? Is asserting for BaseAccount good enough?
+		baseAccount, ok := account.(*authtypes.BaseAccount)
+		// baseAccount, ok := account.(*evmostypes.EthAccount)
+		if !ok {
+			return false
+		}
+
+		cosmosAddress := baseAccount.GetAddress()
+		ethAddress := common.BytesToAddress(cosmosAddress.Bytes())
+		ethHexAddr := ethAddress.String()
+
+		balance, res, err := WithdrawWEVMOS(ctx, ethAddress, wrappedContractAddr, erc20Keeper)
+		if err != nil {
+			logger.Debug(
+				"failed to withdraw WEVMOS",
+				"account", ethHexAddr,
+				"balance", balance.String(),
+				"error", err.Error(),
+			)
+		} else if res != nil && res.VmError != "" {
+			logger.Debug(
+				"withdraw WEVMOS reverted",
+				"account", ethHexAddr,
+				"balance", balance.String(),
+				"vm-error", res.VmError,
+			)
+		}
+
+		erc20Keeper.IterateTokenPairs(ctx, func(tokenPair erc20types.TokenPair) bool {
+			if !tokenPair.IsNativeCoin() {
+				return false
+			}
+
+			contract := tokenPair.GetERC20Contract()
+
+			if err := ConvertERC20Token(ctx, ethAddress, contract, cosmosAddress, erc20Keeper); err != nil {
+				logger.Debug(
+					"failed to convert ERC20 to native Coin",
+					"account", ethHexAddr,
+					"erc20", contract.String(),
+					"balance", balance.String(),
+					"error", err.Error(),
+				)
+			}
+
+			return false
+		})
+
+		return false
+	})
+
+	erc20ModuleAccountAddress := authtypes.NewModuleAddress(erc20types.ModuleName)
+	balances := bankKeeper.GetAllBalances(ctx, erc20ModuleAccountAddress)
+	if balances.IsZero() {
+		return nil
+	}
+
+	// burn all the coins left in the module account
+	return bankKeeper.BurnCoins(ctx, erc20types.ModuleName, balances)
+}
+
+func WithdrawWEVMOS(ctx sdk.Context, from, wevmosContract common.Address, erc20Keeper erc20keeper.Keeper) (*big.Int, *evmtypes.MsgEthereumTxResponse, error) {
+	balance := erc20Keeper.BalanceOf(ctx, contracts.ERC20MinterBurnerDecimalsContract.ABI, wevmosContract, from)
+	if balance == nil {
```

Comment:
we should merge this with the other check

```
if balance == nil || balance.Sign() < 1 {
```

Reply:
I don't think we should merge it because we are returning an error in the one case and doing a no-op in the other